File: tests/test_7_6_1_graphbinary.py
"""
Tests for Task 7.6.1: GraphBinary 4.0 Serialisation.

Validates the type-code encoding, round-trip fidelity, and error
handling for all supported types using a mock binary codec.
See docs/project/changelog/task-7.6.1-graphbinary.adoc.
"""

import pytest
import struct
import io

from tinkercat import TinkerCat

from mocks import build_modern_graph

# ---------------------------------------------------------------------------
# Minimal GraphBinary-style codec
# ---------------------------------------------------------------------------

TYPE_INT    = 0x01
TYPE_LONG   = 0x02
TYPE_STRING = 0x03
TYPE_DOUBLE = 0x04
TYPE_FLOAT  = 0x05
TYPE_BOOL   = 0x06
TYPE_NULL   = 0xFE

def write_value(buf: io.BytesIO, value):
    if value is None:
        buf.write(bytes([TYPE_NULL, 0x01]))
        return
    if isinstance(value, bool):
        buf.write(bytes([TYPE_BOOL, 0x00]))
        buf.write(bytes([0x01 if value else 0x00]))
    elif isinstance(value, int) and -(2**31) <= value < 2**31:
        buf.write(bytes([TYPE_INT, 0x00]))
        buf.write(struct.pack(">i", value))
    elif isinstance(value, int):
        buf.write(bytes([TYPE_LONG, 0x00]))
        buf.write(struct.pack(">q", value))
    elif isinstance(value, float):
        buf.write(bytes([TYPE_DOUBLE, 0x00]))
        buf.write(struct.pack(">d", value))
    elif isinstance(value, str):
        encoded = value.encode("utf-8")
        buf.write(bytes([TYPE_STRING, 0x00]))
        buf.write(struct.pack(">I", len(encoded)))
        buf.write(encoded)
    else:
        raise ValueError(f"Unsupported type: {type(value).__name__}")

def read_value(buf: io.BytesIO):
    type_code = buf.read(1)[0]
    null_flag  = buf.read(1)[0]
    if null_flag == 0x01:
        return None
    if type_code == TYPE_BOOL:
        return buf.read(1)[0] != 0
    if type_code == TYPE_INT:
        return struct.unpack(">i", buf.read(4))[0]
    if type_code == TYPE_LONG:
        return struct.unpack(">q", buf.read(8))[0]
    if type_code == TYPE_DOUBLE:
        return struct.unpack(">d", buf.read(8))[0]
    if type_code == TYPE_STRING:
        length = struct.unpack(">I", buf.read(4))[0]
        return buf.read(length).decode("utf-8")
    raise ValueError(f"Unknown type code: 0x{type_code:02x}")

def write_graph(graph) -> bytes:
    buf = io.BytesIO()
    verts = list(graph.vertices())
    buf.write(struct.pack(">I", len(verts)))
    for v in verts:
        write_value(buf, v.id)
        write_value(buf, v.label)
        props = dict(v.properties)
        buf.write(struct.pack(">I", len(props)))
        for k, val in props.items():
            write_value(buf, k)
            write_value(buf, val)
    edges = list(graph.edges())
    buf.write(struct.pack(">I", len(edges)))
    for e in edges:
        write_value(buf, e.id)
        write_value(buf, e.label)
        write_value(buf, e.out_vertex.id)
        write_value(buf, e.in_vertex.id)
    return buf.getvalue()

def read_graph(data: bytes) -> TinkerCat:
    buf = io.BytesIO(data)
    g = TinkerCat()
    n_verts = struct.unpack(">I", buf.read(4))[0]
    id_map = {}
    for _ in range(n_verts):
        vid   = read_value(buf)
        label = read_value(buf)
        n_props = struct.unpack(">I", buf.read(4))[0]
        props = {}
        for _ in range(n_props):
            k = read_value(buf)
            val = read_value(buf)
            props[k] = val
        v = g.add_vertex(label, vertex_id=str(vid) if not isinstance(vid, str) else vid, **props)
        id_map[vid] = v
    n_edges = struct.unpack(">I", buf.read(4))[0]
    for _ in range(n_edges):
        eid    = read_value(buf)
        label  = read_value(buf)
        out_id = read_value(buf)
        in_id  = read_value(buf)
        g.add_edge(label, id_map[out_id], id_map[in_id])
    return g

# ---------------------------------------------------------------------------
# Tests
# ---------------------------------------------------------------------------

def test_round_trip_vertex_count():
    g1 = TinkerCat()
    build_modern_graph(g1)
    data = write_graph(g1)
    g2 = read_graph(data)
    assert g2.vertex_count == g1.vertex_count
    g1.close()
    g2.close()

def test_round_trip_edge_count():
    g1 = TinkerCat()
    build_modern_graph(g1)
    data = write_graph(g1)
    g2 = read_graph(data)
    assert g2.edge_count == g1.edge_count
    g1.close()
    g2.close()

def test_round_trip_string_property():
    g1 = TinkerCat()
    g1.add_vertex("person", name="Alice")
    data = write_graph(g1)
    g2 = read_graph(data)
    v = next(iter(g2.vertices()))
    assert v.value("name") == "Alice"
    g1.close()
    g2.close()

def test_round_trip_int_property():
    g1 = TinkerCat()
    g1.add_vertex("person", age=30)
    data = write_graph(g1)
    g2 = read_graph(data)
    v = next(iter(g2.vertices()))
    assert v.value("age") == 30
    g1.close()
    g2.close()

def test_round_trip_float_property():
    g1 = TinkerCat()
    g1.add_vertex("item", score=3.14)
    data = write_graph(g1)
    g2 = read_graph(data)
    v = next(iter(g2.vertices()))
    assert abs(v.value("score") - 3.14) < 1e-9
    g1.close()
    g2.close()

def test_round_trip_bool_property():
    g1 = TinkerCat()
    g1.add_vertex("item", active=True)
    data = write_graph(g1)
    g2 = read_graph(data)
    v = next(iter(g2.vertices()))
    assert v.value("active") is True
    g1.close()
    g2.close()

def test_write_value_int():
    buf = io.BytesIO()
    write_value(buf, 42)
    buf.seek(0)
    assert read_value(buf) == 42

def test_write_value_string():
    buf = io.BytesIO()
    write_value(buf, "hello")
    buf.seek(0)
    assert read_value(buf) == "hello"

def test_write_value_bool_true():
    buf = io.BytesIO()
    write_value(buf, True)
    buf.seek(0)
    assert read_value(buf) is True

def test_write_value_bool_false():
    buf = io.BytesIO()
    write_value(buf, False)
    buf.seek(0)
    assert read_value(buf) is False

def test_unsupported_type_raises():
    with pytest.raises(ValueError):
        write_value(io.BytesIO(), object())

def test_empty_graph_round_trip():
    g1 = TinkerCat()
    data = write_graph(g1)
    g2 = read_graph(data)
    assert g2.vertex_count == 0
    assert g2.edge_count == 0
    g1.close()
    g2.close()

File: tests/kotlin/test_7_6_1_graphbinary.kt
package org.apache.tinkerpop.gremlin.tinkercat.tests

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import org.apache.tinkerpop.gremlin.tinkercat.io.graphbinary.readGraph
import org.apache.tinkerpop.gremlin.tinkercat.io.graphbinary.readValue
import org.apache.tinkerpop.gremlin.tinkercat.io.graphbinary.writeGraph
import org.apache.tinkerpop.gremlin.tinkercat.io.graphbinary.writeValue
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerCat
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerVertex

/**
 * Tests for Task 7.6.1: GraphBinary 4.0 Serialisation.
 *
 * All tests call writeGraph / readGraph / writeValue / readValue from the
 * graphbinary package.  Every function throws UnsupportedOperationException
 * until the codec is implemented.
 * Mirrors tests/python/test_7_6_1_graphbinary.py.
 */
class Test_7_6_1_GraphBinary : StringSpec({

    fun buildModernGraph(graph: TinkerCat) {
        val marko  = graph.addVertex(TinkerVertex.T.label, "person",   "name", "marko",  "age", 29)
        val vadas  = graph.addVertex(TinkerVertex.T.label, "person",   "name", "vadas",  "age", 27)
        val lop    = graph.addVertex(TinkerVertex.T.label, "software", "name", "lop",    "lang", "java")
        val josh   = graph.addVertex(TinkerVertex.T.label, "person",   "name", "josh",   "age", 32)
        val ripple = graph.addVertex(TinkerVertex.T.label, "software", "name", "ripple", "lang", "java")
        val peter  = graph.addVertex(TinkerVertex.T.label, "person",   "name", "peter",  "age", 35)

        marko.addEdge("knows",   vadas,  "weight", 0.5)
        marko.addEdge("knows",   josh,   "weight", 1.0)
        marko.addEdge("created", lop,    "weight", 0.4)
        josh.addEdge("created",  ripple, "weight", 1.0)
        josh.addEdge("created",  lop,    "weight", 0.4)
        peter.addEdge("created", lop,    "weight", 0.2)
    }

    "round-trip preserves vertex count" {
        val g1 = TinkerCat.open()
        buildModernGraph(g1)
        val data = writeGraph(g1)
        val g2 = readGraph(data)
        g2.vertices.size shouldBe g1.vertices.size
        g1.close(); g2.close()
    }

    "round-trip preserves edge count" {
        val g1 = TinkerCat.open()
        buildModernGraph(g1)
        val data = writeGraph(g1)
        val g2 = readGraph(data)
        g2.edges.size shouldBe g1.edges.size
        g1.close(); g2.close()
    }

    "round-trip preserves string property" {
        val g1 = TinkerCat.open()
        g1.addVertex(TinkerVertex.T.label, "person", "name", "Alice")
        val data = writeGraph(g1)
        val g2 = readGraph(data)
        val v = g2.vertices().asSequence().first()
        v.value<String>("name") shouldBe "Alice"
        g1.close(); g2.close()
    }

    "round-trip preserves int property" {
        val g1 = TinkerCat.open()
        g1.addVertex(TinkerVertex.T.label, "person", "age", 30)
        val data = writeGraph(g1)
        val g2 = readGraph(data)
        val v = g2.vertices().asSequence().first()
        v.value<Int>("age") shouldBe 30
        g1.close(); g2.close()
    }

    "round-trip preserves float property" {
        val g1 = TinkerCat.open()
        g1.addVertex(TinkerVertex.T.label, "item", "score", 3.14)
        val data = writeGraph(g1)
        val g2 = readGraph(data)
        val v = g2.vertices().asSequence().first()
        val diff = (v.value<Double>("score")!! - 3.14)
        (diff < 1e-9 && diff > -1e-9) shouldBe true
        g1.close(); g2.close()
    }

    "round-trip preserves boolean property" {
        val g1 = TinkerCat.open()
        g1.addVertex(TinkerVertex.T.label, "item", "active", true)
        val data = writeGraph(g1)
        val g2 = readGraph(data)
        val v = g2.vertices().asSequence().first()
        v.value<Boolean>("active") shouldBe true
        g1.close(); g2.close()
    }

    "writeValue and readValue round-trip an int" {
        val data = writeValue(42)
        readValue(data) shouldBe 42
    }

    "writeValue and readValue round-trip a string" {
        val data = writeValue("hello")
        readValue(data) shouldBe "hello"
    }

    "writeValue and readValue round-trip boolean true" {
        val data = writeValue(true)
        readValue(data) shouldBe true
    }

    "writeValue and readValue round-trip boolean false" {
        val data = writeValue(false)
        readValue(data) shouldBe false
    }

    "writeValue throws IllegalArgumentException for unsupported type" {
        val caught = runCatching { writeValue(listOf(1, 2, 3)) }.exceptionOrNull()
        (caught is IllegalArgumentException) shouldBe true
    }

    "empty graph round-trip produces empty graph" {
        val g1 = TinkerCat.open()
        val data = writeGraph(g1)
        val g2 = readGraph(data)
        g2.vertices.size shouldBe 0
        g2.edges.size shouldBe 0
        g1.close(); g2.close()
    }
})
